addons/portal/controllers/portal.py

```diff
@@ -181,10 +181,85 @@ def home(self, **kw):
+            'zip': kw.get('zipcode'),
+            'city': kw.get('city', ''),
+            'state_id': kw.get('state_id'),
+            'type': kw.get('type'),
+        }
+        if address:
+            partner = request.env['res.partner'].sudo().browse(address)
+            country_id = request.env['res.country'].sudo().search([('id', '=', kw.get('country_id', '') or partner.country_id.id)])
+            state_id = request.env['res.country.state'].sudo().search([('id', '=', kw.get('state_id', '') or partner.state_id.id)])
+            address_data.update({
+                'vat': kw.get('vat', '') or partner.vat,
+                'company_name': kw.get('company_name', '') or partner.company_name,
+                'country_id': country_id,
+                'state_id': state_id,
+            })
+            partner.sudo().write(address_data)
+        else:
+            address_data['parent_id'] = request.env.user.partner_id.id
+            address_data.update({'country_id': kw.get('country_id', '') or request.env.user.partner_id.country_id.id})
+            request.env['res.partner'].sudo().create(address_data)
+        return request.redirect('/my/addresses')
+
+    @route(['/my/addresses'], type='http', auth='user', website=True)
+    def addresses(self, **kwargs):
+        values = {
+            **self._prepare_portal_layout_values(),
+            'partner': request.env.user.partner_id,
+            'page_name': 'my_addresses'
+        }
+        # test default
+        if request.env.user.partner_id:
+            if not request.env.user.partner_id.is_default_billing_address and not any(child.is_default_billing_address for child in request.env.user.partner_id.child_ids):
+                request.env.user.partner_id.is_default_billing_address = True
+
+            if not request.env.user.partner_id.is_default_shipping_address and not any(child.is_default_shipping_address for child in request.env.user.partner_id.child_ids):
+                request.env.user.partner_id.is_default_shipping_address = True
+        return request.render("portal.portal_my_addresses", values)
+
+    @route(['/my/addresses/new'], auth='user', website=True)
+    def addresses_new(self, **kw):
+        address_type = kw.get('type')
+        if address_type not in ['invoice', 'delivery']:
+            address_type = 'other'
+        return request.redirect(f"/my/account?type={address_type}")
+
+    @route(['/archive/address/<int:partner_id>'], type='http', methods=['GET', 'POST'], auth="public", website=True, sitemap=False)
+    def address_archive(self, partner_id):
+        if partner_id:
+            partner = request.env['res.partner'].sudo().browse(partner_id)
+            if partner.user_ids.filtered(lambda user: user.active):
+                return request.redirect('/my/addresses?error=True')
+            else:
+                partner.sudo().write({'active': False})
+        return request.redirect('/my/addresses')
```

Comment:
You can archive any partners too as long as they are not linked to active users.